spark-tensorflow/prepare/src/main/scala/com/google/cloud/ml/samples/criteo/CriteoMissingReplacer.scala

```diff
@@ -20,49 +20,53 @@ import org.apache.spark.sql._
 import org.apache.spark.sql.functions.avg
 
 /**
- * Missing replacer replaces all null values in a Datframe with empty strings,
- * and replaces missing integer values with the average of all integer
- * in that column.
- *
- * @param spark Spark session
-**/
+  * Missing replacer replaces all null values in a Datframe with empty strings,
+  * and replaces missing integer values with the average of all integer
+  * in that column.
+  *
+  * @param spark Spark session
+  */
 class CriteoMissingReplacer(val artifactExporter: ArtifactExporter)
                            (implicit val spark: SparkSession) {
   import spark.implicits._
 
-  def preprocessCriteoFeatures(df: DataFrame, integerFeatures: Seq[String]): DataFrame = {
-    val filledDf = df.na.fill("", df.columns)
-
-    averageIntegerFeatures(filledDf, integerFeatures)
-  }
-
-  def averageIntegerFeatures(to_average_df: DataFrame, features: Seq[String]): DataFrame = {
-
+  /**
+    * Calculates a map of integer columns to their average values.
+    * @param to_average_df  The DataFrame with integer columns to get averages of
+    * @param features The column names of the integer features
+    * @return A map from integer column names to their averages
+    */
+  def getAverageIntegerFeatures(to_average_df: DataFrame,
+                                features: Seq[String]): Map[String, DataFrame] = {
     val integerFeaturesDf = to_average_df.
       select(features.head, features.tail: _*).
       toDF
 
-    val averages = integerFeaturesDf.
+    integerFeaturesDf.
```

Comment:
How are null entries treated by this?

One thing that we should be careful about is that the average for a given feature is only taken over cells in the DataFrame which do not have nulls. For this purpose, replacing nulls with 0 would not work.

Reply:
Spark sql average functions just ignores nulls so it does what we want